## 2.- Analsiis de datos extra


In [ ]:
import pandas as pd
import os

In [2]:
data = pd.read_csv(r"..\..\data\raw\train.csv")


NameError: name 'pd' is not defined

In [ ]:
print("\nInformación general:")
data.shape


Información general:


(22835, 21)

In [ ]:
print("VALORES ÚNICOS POR COLUMNA:")
for col in data.columns:
    print(f"\n--- {col} ---")
    unique = data[col].unique()
    print(f'{len(unique)}')
    print(f'{unique}')


📌 VALORES ÚNICOS POR COLUMNA:

--- id ---
22835
['bd9d6267ec5' '9eebf6d8098' '83806858500' ... 'ddb3c5e9bff' '407200a5dfe'
 'eaa66193392']

--- LoanNr_ChkDgt ---
22835
[1523195006 1326365010 6179584001 ... 4082983001 7783283010 1513375008]

--- Name ---
20443
['P-SCAPE LAND DESIGN, LLC' 'The Fresh & Healthy Catering C'
 'AARON MASON & HOWE LLC' ... 'MAINLINE TRCK&TRAILR SRVC, INC'
 'TIN BOX STUDIO' 'Lowex, Inc.']

--- City ---
1679
['NORTHFIELD' 'CANTON' 'SAWYERWOOD' ... 'CASTALIA' 'KITTS HILL'
 'East Canton']

--- State ---
1
['OH']

--- Bank ---
287
['CITIZENS BANK NATL ASSOC' 'FIRSTMERIT BANK, N.A.'
 'PNC BANK, NATIONAL ASSOCIATION' 'THE HUNTINGTON NATIONAL BANK'
 'STEARNS BK NATL ASSOC' 'NATL BK & TR CO' 'WACHOVIA SBA LENDING, INC.'
 'THE CITIZENS SAVINGS BANK' 'ACCESS BUS. DEVEL & FINANCE IN'
 'KEYBANK NATIONAL ASSOCIATION' 'JPMORGAN CHASE BANK NATL ASSOC'
 'THE MERCHANTS BANK & TRUST CO.' 'U.S. BANK NATIONAL ASSOCIATION'
 'FIFTH THIRD BANK' 'CITIZENS BANK' 'CAPITAL ONE NATL ASSOC

In [ ]:
for col in ['State', 'Bank', 'BankState', 'NewExist', 'UrbanRural', 'RevLineCr', 'LowDoc']:
    print(f"\nAceptación media por categoría en '{col}':")
    print(data.groupby(col)['Accept'].mean().sort_values(ascending=False))


📊 Aceptación media por categoría en 'State':
State
OH    0.832231
Name: Accept, dtype: float64

📊 Aceptación media por categoría en 'Bank':
Bank
KEY DIRECT, A BRANCH OF           1.0
PROGRESSIVE BANK NATL ASSOC       1.0
MANUFACTURERS & TRADERS TR CO     1.0
MENTOR ECONOMIC ASSISTANCE COR    1.0
MERCHANTS NATIONAL BANK           1.0
                                 ... 
THE PEOPLES NATIONAL BANK         0.0
BANK SNB, NATIONAL ASSOCIATION    0.0
INEZ DEPOSIT BANK                 0.0
FDIC/FIRST BANK NATIONAL          0.0
KINGSTON NATIONAL BANK            0.0
Name: Accept, Length: 286, dtype: float64

📊 Aceptación media por categoría en 'BankState':
BankState
AL    1.000000
IA    1.000000
WI    1.000000
NM    1.000000
NE    1.000000
AZ    1.000000
ME    1.000000
KS    1.000000
ID    1.000000
WY    1.000000
GA    1.000000
IN    0.983740
TN    0.950000
KY    0.912281
NV    0.900000
PA    0.896825
MN    0.891892
MI    0.891304
OH    0.890555
DC    0.857143
IL    0.834962
WA    0.833333
TX  

In [ ]:
# 1. Filtrar solo préstamos aprobados
aprobados = data[data['Accept'] == 1].copy()

print(f"Total préstamos aprobados: {len(aprobados)}")

# 2. Ver valores únicos por columna en el nuevo dataset
print("\nValores únicos por columna en préstamos aprobados:\n")
for col in aprobados.columns:
    print(f"\n--- {col} ---")
    unique = aprobados[col].unique()
    print(f'{len(unique)}')
    print(f'{unique}')


✅ Total préstamos aprobados: 19004

🔍 Valores únicos por columna en préstamos aprobados:


--- id ---
19004
['bd9d6267ec5' '9eebf6d8098' '83806858500' ... '798db2753a7' 'ddb3c5e9bff'
 '407200a5dfe']

--- LoanNr_ChkDgt ---
19004
[1523195006 1326365010 6179584001 ... 2011184008 4082983001 7783283010]

--- Name ---
17107
['P-SCAPE LAND DESIGN, LLC' 'The Fresh & Healthy Catering C'
 'AARON MASON & HOWE LLC' ... 'ENVIRO SHIELD POWER WASHING'
 'MAINLINE TRCK&TRAILR SRVC, INC' 'TIN BOX STUDIO']

--- City ---
1549
['NORTHFIELD' 'CANTON' 'SAWYERWOOD' ... 'Northfield Center' 'KITTS HILL'
 'East Canton']

--- State ---
1
['OH']

--- Bank ---
262
['CITIZENS BANK NATL ASSOC' 'FIRSTMERIT BANK, N.A.'
 'PNC BANK, NATIONAL ASSOCIATION' 'THE HUNTINGTON NATIONAL BANK'
 'STEARNS BK NATL ASSOC' 'NATL BK & TR CO' 'WACHOVIA SBA LENDING, INC.'
 'ACCESS BUS. DEVEL & FINANCE IN' 'KEYBANK NATIONAL ASSOCIATION'
 'JPMORGAN CHASE BANK NATL ASSOC' 'THE MERCHANTS BANK & TRUST CO.'
 'U.S. BANK NATIONAL ASSOCIATION' 'F

In [ ]:
def parse_currency(value):
    if pd.isna(value):
        return None
    return float(str(value).replace('$', '').replace(',', ''))
data.head()
data['DisbursementGross'] = data['DisbursementGross'].apply(parse_currency)
data.head()

id  LoanNr_ChkDgt                            Name        City  \
0  bd9d6267ec5     1523195006        P-SCAPE LAND DESIGN, LLC  NORTHFIELD   
1  9eebf6d8098     1326365010  The Fresh & Healthy Catering C      CANTON   
2  83806858500     6179584001          AARON MASON & HOWE LLC  SAWYERWOOD   
3  a21ab9cb3af     8463493009               MID OHIO CAR WASH    COLUMBUS   
4  883b5e5385e     3382225007                 Bake N Brew LLC      Newark   

  State                            Bank BankState ApprovalDate  ApprovalFY  \
0    OH        CITIZENS BANK NATL ASSOC        RI     1-Nov-05        2006   
1    OH           FIRSTMERIT BANK, N.A.        OH     6-Jun-05        2005   
2    OH  PNC BANK, NATIONAL ASSOCIATION        OH    18-Mar-03        2003   
3    OH    THE HUNTINGTON NATIONAL BANK        OH    28-Jun-95        1995   
4    OH    THE HUNTINGTON NATIONAL BANK        OH    16-Apr-09        2009   

   NoEmp  ...  CreateJob  RetainedJob  FranchiseCode  UrbanRural  RevLineCr  \
0      2  ...          0            2              0           1          N   
1      2  ...          1            2              1           1          N   
2      2  ...          4            2              1           2          Y   
3      2  ...          0            0              1           0          N   
4      0  ...          0            0              0           1          N   

  LowDoc DisbursementDate DisbursementGross  BalanceGross Accept  
0      N        31-Dec-05            8000.0        $0.00       1  
1      N        31-Jul-05          166000.0        $0.00       1  
2      N        31-Mar-03           25000.0        $0.00       1  
3      N        31-Jan-96          220100.0        $0.00       1  
4      N        31-May-09           25000.0        $0.00       0  

[5 rows x 21 columns]

In [ ]:
data['ApprovalDate'] = pd.to_datetime(data['ApprovalDate'], errors='coerce')
data['DisbursementDate'] = pd.to_datetime(data['DisbursementDate'], errors='coerce')
data['FranchiseCode'] = data['FranchiseCode'].astype(str)
data['UrbanRural'] = data['UrbanRural'].astype(str)
data['NewExist'] = data['NewExist'].astype(str)
data['ApprovalFY'] = data['ApprovalFY'].astype(str)

print("Nulos por columna tras limpieza inicial:")
print(data.isnull().sum())

# 7. Visualizar tipos de datos actuales
print("\nTipos de datos tras conversión:")
print(data.dtypes)

📌 Nulos por columna tras limpieza inicial:
id                    0
LoanNr_ChkDgt         0
Name                  1
City                  1
State                 0
Bank                 22
BankState            22
ApprovalDate          0
ApprovalFY            0
NoEmp                 0
NewExist              0
CreateJob             0
RetainedJob           0
FranchiseCode         0
UrbanRural            0
RevLineCr            91
LowDoc               43
DisbursementDate     61
DisbursementGross     0
BalanceGross          0
Accept                0
dtype: int64

🧾 Tipos de datos tras conversión:
id                           object
LoanNr_ChkDgt                 int64
Name                         object
City                         object
State                        object
Bank                         object
BankState                    object
ApprovalDate         datetime64[ns]
ApprovalFY                   object
NoEmp                         int64
NewExist                     object
CreateJob

In [ ]:
data['RevLineCr'] = data['RevLineCr'].replace({
    '0': 'N',
    'T': 'Y',
    '`': 'NA'  # o np.nan si prefieres imputar después
})
data['RevLineCr'] = data['RevLineCr'].fillna('NA')

In [ ]:
print(data['RevLineCr'].unique())

['N' 'Y' 'NA']


In [ ]:
print(data['RevLineCr'].value_counts(dropna=False))

RevLineCr
N     16353
Y      6390
NA       92
Name: count, dtype: int64


In [ ]:
# Limpiar valores incorrectos en LowDoc
data['LowDoc'] = data['LowDoc'].replace({
    '0': 'N',
    'A': 'NA',
    'S': 'NA',
    'R': 'NA',
    'C': 'NA'
})

# Rellenar los NaN restantes con 'NA'
data['LowDoc'] = data['LowDoc'].fillna('NA')

# Confirmar limpieza
print("\nValores únicos en LowDoc después de limpiar:")
print(data['LowDoc'].unique())



✅ Valores únicos en LowDoc después de limpiar:
['N' 'Y' 'NA']


In [ ]:
print(data['LowDoc'].value_counts(dropna=False))

LowDoc
N     19681
Y      3094
NA       60
Name: count, dtype: int64


In [ ]:
data['NewExist'] = data['NewExist'].replace({
    '0.0': 'NA',
    'nan': 'NA' 
})
# Confirmar limpieza
print("\n Valores únicos en NewExist después de limpiar:")
print(data['NewExist'].unique())


 Valores únicos en NewExist después de limpiar:
['2.0' '1.0' 'NA']


In [ ]:
print(data['UrbanRural'].value_counts(dropna=False))

UrbanRural
1    11582
0     7538
2     3715
Name: count, dtype: int64


In [ ]:
data['IsFranchise'] = data['FranchiseCode'].apply(lambda x: 'No' if x in ['0', '1'] else 'Sí')


In [ ]:
print(data['IsFranchise'].value_counts(dropna=False))

IsFranchise
No    21267
Sí     1568
Name: count, dtype: int64


In [ ]:
print("Estadísticas de columnas numéricas:")
print(data[['NoEmp', 'CreateJob', 'RetainedJob', 'DisbursementGross']].describe())

📊 Estadísticas de columnas numéricas:
              NoEmp     CreateJob   RetainedJob  DisbursementGross
count  22835.000000  22835.000000  22835.000000       2.283500e+04
mean      12.368995     18.479571     20.960981       1.693213e+05
std       68.142670    377.130205    377.219172       2.491081e+05
min        0.000000      0.000000      0.000000       0.000000e+00
25%        2.000000      0.000000      0.000000       4.000000e+04
50%        5.000000      0.000000      0.000000       8.750000e+04
75%       12.000000      2.000000      4.000000       1.891895e+05
max     7538.000000   8800.000000   8800.000000       4.831510e+06


In [ ]:
print(" Total de préstamos con 0 empleados:")
print((data['NoEmp'] == 0).sum())

# ¿Cuál es su proporción de aprobación?
print("\nTasa de aprobación para empresas con 0 empleados:")
print(data[data['NoEmp'] == 0]['Accept'].mean())

🔍 Total de préstamos con 0 empleados:
994

📈 Tasa de aprobación para empresas con 0 empleados:
0.8078470824949698


In [ ]:
print(" VALORES ÚNICOS POR COLUMNA:")
for col in data.columns:
    print(f"\n--- {col} ---")
    unique = data[col].unique()
    print(f'{len(unique)}')
    print(f'{unique}')


📌 VALORES ÚNICOS POR COLUMNA:

--- id ---
22835
['bd9d6267ec5' '9eebf6d8098' '83806858500' ... 'ddb3c5e9bff' '407200a5dfe'
 'eaa66193392']

--- LoanNr_ChkDgt ---
22835
[1523195006 1326365010 6179584001 ... 4082983001 7783283010 1513375008]

--- Name ---
20443
['P-SCAPE LAND DESIGN, LLC' 'The Fresh & Healthy Catering C'
 'AARON MASON & HOWE LLC' ... 'MAINLINE TRCK&TRAILR SRVC, INC'
 'TIN BOX STUDIO' 'Lowex, Inc.']

--- City ---
1679
['NORTHFIELD' 'CANTON' 'SAWYERWOOD' ... 'CASTALIA' 'KITTS HILL'
 'East Canton']

--- State ---
1
['OH']

--- Bank ---
287
['CITIZENS BANK NATL ASSOC' 'FIRSTMERIT BANK, N.A.'
 'PNC BANK, NATIONAL ASSOCIATION' 'THE HUNTINGTON NATIONAL BANK'
 'STEARNS BK NATL ASSOC' 'NATL BK & TR CO' 'WACHOVIA SBA LENDING, INC.'
 'THE CITIZENS SAVINGS BANK' 'ACCESS BUS. DEVEL & FINANCE IN'
 'KEYBANK NATIONAL ASSOCIATION' 'JPMORGAN CHASE BANK NATL ASSOC'
 'THE MERCHANTS BANK & TRUST CO.' 'U.S. BANK NATIONAL ASSOCIATION'
 'FIFTH THIRD BANK' 'CITIZENS BANK' 'CAPITAL ONE NATL ASSOC

In [ ]:
inconsistentes_create = data[data['RetainedJob'] > (data['NoEmp'] + data['CreateJob'])] 
print(f" Casos donde se crean más empleos de los que tiene la empresa: {len(inconsistentes_create)}")

# Mostrar algunos
print(inconsistentes_create[['NoEmp', 'CreateJob', 'RetainedJob', 'Accept']])

print("\n Distribución en inconsistencias:")
print(inconsistentes_create['Accept'].value_counts(normalize=True))

⚠️ Casos donde se crean más empleos de los que tiene la empresa: 111
       NoEmp  CreateJob  RetainedJob  Accept
102        0          3            5       0
270        7          0            9       0
406        2          3           35       1
656        1          0            7       1
657        0          0            1       0
...      ...        ...          ...     ...
22269      0          0            3       1
22461      0          2            3       1
22466      2          0            8       1
22532      0          0            1       1
22762      8          2           11       1

[111 rows x 4 columns]

📊 Distribución en inconsistencias:
Accept
1    0.711712
0    0.288288
Name: proportion, dtype: float64


In [1]:
data['ApprovalDate'] = pd.to_datetime(data['ApprovalDate'], errors='coerce')
data['DisbursementDate'] = pd.to_datetime(data['DisbursementDate'], errors='coerce')


NameError: name 'pd' is not defined

In [ ]:
data.to_csv("train_limpio.csv", index=False)


In [4]:
import pandas as pd

df = pd.read_csv('../3_modelado/3.3_modelo_secreto/train_limpio.csv')

# 1. Eliminar columnas irrelevantes
cols_to_drop = ['id', 'LoanNr_ChkDgt', 'Name']
df.drop(columns=cols_to_drop, inplace=True)

# 2. Reemplazar todos los NaNs por 'Unknown', sin suponer valores
df = df.fillna('Unknown')

# 3. Conversión de fechas: convertir pero mantener Unknown si no se puede parsear
for date_col in ['ApprovalDate', 'DisbursementDate']:
    df[date_col] = pd.to_datetime(df[date_col], errors='coerce')
    df[date_col] = df[date_col].fillna('Unknown')

# 4. Extraer año y mes si la fecha es válida, si no poner 'Unknown'
df['ApprovalYear'] = df['ApprovalDate'].apply(lambda x: x.year if x != 'Unknown' else 'Unknown')
df['ApprovalMonth'] = df['ApprovalDate'].apply(lambda x: x.month if x != 'Unknown' else 'Unknown')

# 5. Limpiar y convertir BalanceGross (monetario)
df['BalanceGross'] = df['BalanceGross'].astype(str).replace('[\$,]', '', regex=True)
df['BalanceGross'] = df['BalanceGross'].apply(lambda x: float(x) if x.replace('.', '', 1).isdigit() else 'Unknown')

# 6. Mapear binarios si son Y/N o Yes/No, si no están especificados → '2'
def map_binary(val):
    if val in ['Y', 'Yes']:
        return 1
    elif val in ['N', 'No']:
        return 0
    else:
        return 2

df['RevLineCr'] = df['RevLineCr'].apply(map_binary)
df['LowDoc'] = df['LowDoc'].apply(map_binary)
df['IsFranchise'] = df['IsFranchise'].apply(map_binary)

# 7. Convertir variables categóricas (CatBoost las puede manejar como 'category')
categorical_columns = ['City', 'State', 'Bank', 'BankState']
for col in categorical_columns:
    df[col] = df[col].astype(str) 

In [5]:
df.to_csv('train_limpio_v2.csv')

## Conclusiones

### Limpieza de columnas categóricas

RevLineCr y LowDoc contenían valores anómalos como '0', 'T', o símbolos ('), que fueron corregidos y estandarizados a 'Y', 'N', o 'NA'.

En LowDoc se detectaron valores erróneos ('A', 'R', 'S', 'C') que fueron considerados como información faltante y recodificados como 'NA'.

Se creó una nueva columna IsFranchise, que binariza FranchiseCode: 'No' si es '0' o '1', 'Sí' en caso contrario

## Analisis de columnas numéricas

994 empresas tienen NoEmp = 0, de las cuales muchas están aprobadas (tasa de aprobación ≈ 0.8078470824949698).
1555 registros tenían más empleos creados (CreateJob) que empleados actuales (NoEmp)  
111 registros tenían RetainedJob > NoEmp + CreateJob (tasa de aprobación ≈ 0.711712)